src/openassetio-core/src/pluginSystem/HybridPluginSystemManagerImplementationFactory.cpp

```diff
@@ -0,0 +1,421 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 The Foundry Visionmongers Ltd
+#include <algorithm>
+#include <cassert>
+#include <memory>
+
+#include <fmt/format.h>
+
+#include <openassetio/errors/exceptions.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/pluginSystem/HybridPluginSystemManagerImplementationFactory.hpp>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace pluginSystem {
+
+namespace {
+
+/**
+ * A ManagerInterface implementation that composes multiple child
+ * ManagerInterfaces and forwards API calls to one of the children.
+ *
+ * For most API methods, the call is forwarded to the first child that
+ * satisfies the required capability for that API call, or calls the
+ * base class implementation if no child satisfies the required
+ * capability. The priorirty order of children is defined by the order
+ * that they were provided to the constructor.
+ *
+ * For API calls that have no associated capability, either the first
+ * child is chosen, or the results from all children are merged - see
+ * method-specific docs for details.
+ */
+class HybridManagerInterface : public managerApi::ManagerInterface {
+  using ManagerInterfaces = std::vector<managerApi::ManagerInterfacePtr>;
+
+ public:
+  explicit HybridManagerInterface(ManagerInterfaces managerInterfacess)
+      : managerInterfaces_{std::move(managerInterfacess)} {
+    // Precondition.
+    assert(!managerInterfaces_.empty());
+
+    // Cache a mapping of the first child interface that supports each
+    // capability.
+    //
+    // Caching now avoids the need to call hasCapability() on
+    // the child implementations again (and repeatedly) in the API
+    // methods, which could be expensive if, for example, a network call
+    // is required, or we are calling out to Python and locking the
+    // GIL, etc. The disadvantage is that capabilities cannot change
+    // after plugins have been loaded.
+    managerInterfacesByCapability_.reserve(kCapabilityNames.size());
+    for (std::size_t capabilityIdx = 0; capabilityIdx < kCapabilityNames.size(); ++capabilityIdx) {
+      const auto capability = static_cast<Capability>(capabilityIdx);
+      for (const auto& managerInterface : managerInterfaces_) {
+        if (managerInterface->hasCapability(capability)) {
+          managerInterfacesByCapability_[capability] = managerInterface;
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Return the identifier of the first child, on the understanding that
+   * all child implementations have the same identifier.
```

Comment:
Perhaps. We know that the child plugins advertise the same identifier, otherwise they wouldn't have been composed. 

We require that the `ManagerInterface` instance created by a plugin advertises exactly the same identifier as the plugin (this is checked by the - optional - API Compliance test suite `apiCompliance.py`, but that's a whole other topic).

We don't currently validate that the composite `ManagerInterface` instances of the hybrid plugin definitely advertise the same identifier. But it would be a poorly configured plugin if that were the case, since it would break our requirement.

I guess there's

Reply:
And just like that, you've made me realise another bug - well, missing feature. The API Compliance suite won't work with hybrid plugins!